src/ledger/LedgerEntryScope.h

```diff
@@ -0,0 +1,280 @@
+#pragma once
+
+// Copyright 2025 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "xdr/Stellar-ledger-entries.h"
+#include <cstdint>
+#include <iosfwd>
+#include <optional>
+
+// Defines the static set of scopes. Each of these turns into a static enum
+// entry in StaticLedgerEntryScope as well as a bunch of type synonyms for
+// scoped ledger entries parameterized by the enums.
+
+#define FOREACH_STATIC_LEDGER_ENTRY_SCOPE(MACRO) \
+    MACRO(GlobalParApply) \
+    MACRO(ThreadParApply) \
+    MACRO(TxParApply) \
+    MACRO(LclSnapshot) \
+    MACRO(HotArchive) \
+    MACRO(RawBucket)
+
+// Defines valid scope transitions as (DEST, SOURCE) pairs for compile-time
+// enforcement. The "adopt" methods only allow ScopedLedgerEntries to
+// be adopted between pairs of scopes listed here.
+#define FOR_EACH_VALID_SCOPE_ADOPTION(MACRO) \
+    MACRO(GlobalParApply, ThreadParApply) \
+    MACRO(ThreadParApply, GlobalParApply) \
+    MACRO(ThreadParApply, TxParApply) \
+    MACRO(TxParApply, ThreadParApply)
+
+namespace stellar
+{
+
+enum class StaticLedgerEntryScope : uint16_t
+{
+#define STATIC_SCOPE_MACRO(SCOPE_NAME) SCOPE_NAME,
+    FOREACH_STATIC_LEDGER_ENTRY_SCOPE(STATIC_SCOPE_MACRO)
+#undef STATIC_SCOPE_MACRO
+};
+
+template <StaticLedgerEntryScope> class LedgerEntryScopeID;
+
+}
+
+template <stellar::StaticLedgerEntryScope T>
+std::ostream& operator<<(std::ostream& os,
+                         stellar::LedgerEntryScopeID<T> const& obj);
+
+namespace stellar
+{
+
+// LedgerEntryScopeID uniquely identifies a scope for ledger entries. It is
+// space-optimized to a single 64 bit word since these are stuck on all
+// ScopedLedgerEntry instances and there might be a lot of those.
+template <StaticLedgerEntryScope S> class LedgerEntryScopeID
+{
+  public:
+    static constexpr StaticLedgerEntryScope staticScope = S;
+    uint16_t const mIndex;
+    uint32_t const mLedger;
+
+    // NB: index is limited to 16 bits to keep size down;
+    // the constructor will throw if you exceed that, but
+    // in practice scopes should always be tiny numbers like
+    // the thread cluster index or the bucketlist level.
+    LedgerEntryScopeID(size_t index, uint32_t ledger);
```

Comment:
I'm not sure I understand why don't we pass uint16_t here explicitly.

Reply:
the number usually comes from a size_t (like the index of an array) and if we had callers just _cast_ the size_t to uint16_t they would potentially not notice a truncation. whereas if we centralize the code to check for truncation here we can have better odds of catching one.